Fixed Showing no validation errors when form opens

aurangzaibumer opened the pull request:
**Fixes**
#1001 #https://github.com/opensrp/fhircore/issues/873

**Description**
Removed validation from being triggered when user opens the form and enters no data.

**Type**
Choose one: ( Bug fix)

**Screenshots (if applicable)**
https://user-images.githubusercontent.com/35099184/152525877-d7d52896-3063-4925-903e-159a34d86a32.mp4

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

joiskash commented:
Can you please update the PR template. You can refer to other PRs to know how to do this. eg: https://github.com/google/android-fhir/pull/1106

aurangzaibumer commented:
> Can you please update the PR template. You can refer to other PRs to know how to do this. eg: #1106

thanks, just updated the template.

jingtang10 commented:
Please take a look at this section of the wiki: https://github.com/google/android-fhir/wiki/Structured-Data-Capture-Library#provide-configuration

jingtang10 commented:
discussed offline with @aurangzaibumer: @ellykits and @aurangzaibumer please consider using the `DataCaptureConfg` class. Application developers should be able to set the config using the link I posted above.

If that is indeed a problem, then please let us know so we can make sure the APIs are easy to use.

Just a final note, this way of providing library config is consistent with other android libraries such as work manager.

ellykits commented:
@jingtang10 This is well noted. There was no problem using the `DataCaptureConfg` class. However, one of the potential issues we may  encounter in FHIRCore (currently using Hilt for Dependency Injection) is in the tests requiring the application to be cast into `DataCaptureConfig.Provider` instance. Hilt provides `HiltTestApplication` that is recommended to be used in the instrumentation tests instead of a custom test application class. 

**Note** that it is possible to provide a custom application for Hilt tests however for best practices it is not recommended to avoid accidental leaks of state used in the application class, that is documented here [Hilt custom test application](https://dagger.dev/hilt/testing.html#custom-test-application).

As an alternative approach to implementing the `DataCaptureConfig.Provider` interface via the `Application` class we can consider providing the `DataCaptureConfig` as a singleton initialized in the `onCreate` method of the `Application`. Implementing a custom Hilt test application for our tests would also still address our concern.

jingtang10 commented:
> @jingtang10 This is well noted. There was no problem using the `DataCaptureConfg` class. However, one potential issue we may encounter in FHIRCore (currently using Hilt for Dependency Injection) is in the tests that would require the application to be cast into `DataCaptureConfig.Provider` instance. Hilt provides `HiltTestApplication` that is recommended to be used in the instrumentation tests instead of a custom test application class.
> 
> **Note** that it is possible to provide a custom application for Hilt tests however for best practices it is not recommended to avoid accidental leaks of state used in the application class, that is documented here [Hilt custom test application](https://dagger.dev/hilt/testing.html#custom-test-application).
> 
> As an alternative approach to implementing the `DataCaptureConfig.Provider` interface via the `Application` class we can consider providing the `DataCaptureConfig` through a singleton initialized in the `onCreate` method of the `Application`. Implementing a custom Hilt test application for our tests would also still address our concern.

thanks @ellykits - i don't really understand the reasoning behind Hilt's recommendation besides "it's easy to get it wrong" it seems a bit weak to me. @yigit @stevenckngaa @kevinmost fyi. let me do some research and get back to you on this one.

kevinmost commented:
> As a best practice, avoid using @CustomTestApplication and instead use HiltTestApplication in your tests. In general, having your Activity, Fragment, etc. be independent of the parent they are contained in makes it easier to compose and reuse it in the future.

Sounds like the main concern is that it introduces a stronger coupling between your Hilt Application and the component (Fragment, Activity, etc)?

jingtang10 commented:
i chatted with someone working on hilt and they said: `It is just a 'recommendation', we know it might not be possible for all situations to only use the HiltTestApplication which is why there is also an API for custom applications.`

ellykits commented:
Cool thanks.

On Wed, Feb 16, 2022, 12:30 Jing Tang ***@***.***> wrote:

> i chatted with someone working on hilt and they said: It is just a
> 'recommendation', we know it might not be possible for all situations to
> only use the HiltTestApplication which is why there is also an API for
> custom applications.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/google/android-fhir/pull/1102#issuecomment-1041288391>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGM667BL4B4OUKECQOB4CDLU3NVD5ANCNFSM5NLRF7UA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewHolderFactory.kt:
```diff
@@ -62,7 +63,9 @@ open class QuestionnaireItemViewHolder(
     delegate.bind(questionnaireItemViewItem)
     delegate.setReadOnly(questionnaireItemViewItem.questionnaireItem.readOnly)
-    delegate.displayValidationResult(delegate.getValidationResult(itemView.context))
+    if (delegate.shouldValidateOnStart(itemView.context)) {
+      delegate.displayValidationResult(delegate.getValidationResult(itemView.context))
+    }
```
if you scroll pass the item so that the item disappears, and scroll back so the item reappears, the error message would disappear in this case. this is incorrect.

there needs to be some kind of state attached to the questionnaire item so you can avoid displaying validation result the first time...

jingtang10 commented:
as discussed on friday please see #1209 for a suggested fix.

there're a few issues related to this:

1. we need to unset the on click listener when setting the answer programatically to the UI
2. the `getValue` function in `QuestionnaireItemEditTextStringViewHolderDelegate.kt` should always return a non-null value. This is because there will always be a `QuestionnaireResponseItem` even if it doesn't have an answer.
3. Empty string should be considered by the required validator as an invalid answer.

aurangzaibumer commented:
> as discussed on friday please see #1209 for a suggested fix.
> 
> there're a few issues related to this:
> 
> 1. we need to unset the on click listener when setting the answer programatically to the UI
> 2. the `getValue` function in `QuestionnaireItemEditTextStringViewHolderDelegate.kt` should always return a non-null value. This is because there will always be a `QuestionnaireResponseItem` even if it doesn't have an answer.
> 3. Empty string should be considered by the required validator as an invalid answer.

Thanks Jing
As discussed on Friday, added other types check on Validator. Please share feedback.

fun QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent.isNotValid(): Boolean =
  when {
    hasValueStringType() && valueStringType.value.isNullOrEmpty() -> true
    hasValueBooleanType() && valueBooleanType.isEmpty -> true
    hasValueDecimalType() && valueDecimalType.isEmpty -> true
    hasValueDateType() && valueDateType.isEmpty -> true
    hasValueDateTimeType() && valueDateTimeType.isEmpty -> true
    hasValueQuantity() && valueQuantity.isEmpty -> true
    hasValueCoding() && valueCoding.isEmpty -> true
    hasValue() && value.isEmpty -> true
    hasValueUriType() && valueUriType.isEmpty -> true
    hasValueAttachment() && valueAttachment.isEmpty -> true
    else -> false
  }

aurangzaibumer commented:
Hello @jingtang10 

I locally compared the Jacoco report of the master branch with that of my branch and verified that the lines of code that dropped in coverage were not part of the changes I made. Thanks 

cc @f-odhiambo @maimoonak

Tarun-Bhardwaj commented:
@f-odhiambo , has this PR been internally reviewed and ready for review by Google?

maimoonak commented:
@Tarun-Bhardwaj yes. this has been reviewed.

f-odhiambo commented:
@Tarun-Bhardwaj any feedback on this?

jingtang10 commented on buildSrc/src/main/kotlin/Dependencies.kt:
```diff
@@ -131,6 +131,8 @@ object Dependencies {
   }
 
   const val androidJunitRunner = "androidx.test.runner.AndroidJUnitRunner"
+  const val androidDataCaptureTestRunner =
+    "com.google.android.fhir.datacapture.DataCaptureTestRunner"
```
do you still need this and the changes in the configuration? seems to me they can all be reverted?

jingtang10 commented on datacapture/src/androidTest/java/com/google/android/fhir/datacapture/DataCaptureInstrumentationTestApplication.kt:
```diff
@@ -0,0 +1,28 @@
+/*
```
do you still need this class?

jingtang10 commented on datacapture/src/androidTest/java/com/google/android/fhir/datacapture/DataCaptureTestRunner.kt:
```diff
@@ -0,0 +1,36 @@
+/*
```
and this

jingtang10 commented on datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemAutoCompleteViewHolderFactoryInstrumentedTest.kt:
```diff
@@ -207,7 +207,7 @@ class QuestionnaireItemAutoCompleteViewHolderFactoryInstrumentedTest {
   @Test
   @UiThreadTest
-  fun displayValidationResult_error_shouldShowErrorMessage() {
+  fun displayValidationResult_noError_shouldShowNoErrorMessageAtStart() {
```
you need to add a new test cases. there'll be two different scenarios. 1) questionnaire response without an answer 2) questionnaire response with an answer which has no value

the behavior should be different.

jingtang10 replied:
is this addressed?

aurangzaibumer replied:
```
@Test
  @UiThreadTest
  fun displayValidationResult_questionnaireResponseAnswerWithNoValue_shouldShowErrorMessage() {
    viewHolder.bind(
      QuestionnaireItemViewItem(
        Questionnaire.QuestionnaireItemComponent().apply { required = true
          addAnswerOption(
            Questionnaire.QuestionnaireItemAnswerOptionComponent().apply {
              value =  StringType("hello")
            }
          )},
        QuestionnaireResponse.QuestionnaireResponseItemComponent().apply {
          addAnswer()
        }
      ) {}
    )

    assertThat(viewHolder.itemView.findViewById<TextInputLayout>(R.id.text_input_layout).error)
      .isEqualTo(viewHolder.itemView.context.getString(R.string.required_constraint_validation_error_msg))
  }


  @Test
  @UiThreadTest
  fun displayValidationResult_noQuestionnaireResponseAnswer_shouldShowNoErrorMessage() {
    viewHolder.bind(
      QuestionnaireItemViewItem(
        Questionnaire.QuestionnaireItemComponent().apply { required = true
          addAnswerOption(
            Questionnaire.QuestionnaireItemAnswerOptionComponent().apply {
              value = StringType("hello")
            }
          )},
        QuestionnaireResponse.QuestionnaireResponseItemComponent()
      ) {}
    )

    assertThat(viewHolder.itemView.findViewById<TextInputLayout>(R.id.text_input_layout).error)
      .isEqualTo(null)
  }
```

I've added two cases just like you mentioned for class QuestionnaireItemEditTextSingleLineViewHolderFactoryInstrumentedTest, can you please review.

jingtang10 commented on datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemBooleanTypePickerViewHolderFactoryInstrumentedTest.kt:
```diff
@@ -322,16 +322,15 @@ class QuestionnaireItemBooleanTypePickerViewHolderFactoryInstrumentedTest {
   @Test
   @UiThreadTest
-  fun displayValidationResult_error_shouldShowErrorMessage() {
+  fun displayValidationResult_noError_shouldShowNoErrorMessageAtStart() {
```
same here

aurangzaibumer replied:
as per investigation, Pickerview/Radio button can have only two states
1 -> no QR answer
2 -> QR answer with a value
have added following test cases, can you please review

```
@Test
  @UiThreadTest
  fun displayValidationResult_questionnaireResponseAnswerWithValue_shouldShowNoErrorMessage() {
    viewHolder.bind(
      QuestionnaireItemViewItem(
        Questionnaire.QuestionnaireItemComponent().apply { required = true
          addAnswerOption(
            Questionnaire.QuestionnaireItemAnswerOptionComponent().apply {
              text = "Question?"
            }
          )},
        QuestionnaireResponse.QuestionnaireResponseItemComponent().apply {
          addAnswer(
            QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
              value = BooleanType(false)
            }
          )
        }
      ) {}
    )

    assertThat(viewHolder.itemView.findViewById<TextView>(R.id.error_text_view).error)
      .isEqualTo(null)
  }

  @Test
  @UiThreadTest
  fun displayValidationResult_noQuestionnaireResponseAnswer_shouldShowNoErrorMessage() {
    viewHolder.bind(
      QuestionnaireItemViewItem(
        Questionnaire.QuestionnaireItemComponent().apply { required = true
          addAnswerOption(
            Questionnaire.QuestionnaireItemAnswerOptionComponent().apply {
              text = "Question?"
            }
          )},
        QuestionnaireResponse.QuestionnaireResponseItemComponent()
      ) {}
    )

    assertThat(viewHolder.itemView.findViewById<TextView>(R.id.error_text_view).error)
      .isEqualTo(null)
  }
```

aurangzaibumer commented:
As per discussion with @jingtang10  on 18 March 2022

We've decided to keep a flag on Validation API to work differently whenever submit button is pressed. We will incorporate this change on separate PR. Thanks

cc @f-odhiambo

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewHolderFactory.kt:
```diff
@@ -62,7 +62,9 @@ open class QuestionnaireItemViewHolder(
     delegate.bind(questionnaireItemViewItem)
     delegate.setReadOnly(questionnaireItemViewItem.questionnaireItem.readOnly)
-    delegate.displayValidationResult(delegate.getValidationResult(itemView.context))
+    if (delegate.questionnaireItemViewItem.questionnaireResponseItem.hasAnswer()) {
```
Add some comment here to explain why:

```suggestion
    // Only validate questionnaire items with answer(s). This is so that we do not show all the validation errors at once when the user opens a new questionnaire for the first time. Instead, the validation errors are shown when the user goes through each question.
    // Notice the difference between a questionnnaire response item without answer, and a questionnaire with an answer without value. 
    if (delegate.questionnaireItemViewItem.questionnaireResponseItem.hasAnswer()) {
```

Tarun-Bhardwaj commented:
@aurangzaibumer , have the last few comments from @jingtang10 addressed?
CC: @f-odhiambo

aurangzaibumer commented:
> @aurangzaibumer , have the last few comments from @jingtang10 addressed? CC: @f-odhiambo

I've addressed the comments. Waiting for a feedback. thanks.
cc @f-odhiambo

aurangzaibumer commented:
@jingtang10 can you please review and let me know any feedback. thanks

cc @f-odhiambo

aurangzaibumer commented on datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemEditTextSingleLineViewHolderFactoryInstrumentedTest.kt:
```diff
@@ -205,6 +209,50 @@ class QuestionnaireItemEditTextSingleLineViewHolderFactoryInstrumentedTest {
+
+  @Test
+  @UiThreadTest
+  fun displayValidationResult_noQuestionnaireResponseAnswer_shouldShowNoErrorMessage() {
```
resolved

aurangzaibumer replied:
added new test case for QuestionnaireItemAutoCompleteViewHolderFactoryInstrumentedTest 
can you please review
cc @f-odhiambo

aurangzaibumer commented on datacapture/src/main/java/com/google/android/fhir/datacapture/validation/RequiredConstraintValidator.kt:
```diff
@@ -33,6 +33,14 @@ internal object RequiredConstraintValidator : ConstraintValidator {
+          context.getString(R.string.required_constraint_validation_error_msg)
+        )
+      }
+    }
```
resolved as per the feedback, please review. thanks

aurangzaibumer commented:
> Can you add a test case for `RequiredConstraintValidator.kt`?
> 
> Actually if you do this, you can probably remove the new test cases you added in the boolean and single line widgets' tests.

I've added few test cases for the `RequiredConstraintValidator.kt` class as per the business. Can you please review. thanks

jingtang10 commented on datacapture/src/test/java/com/google/android/fhir/datacapture/validation/RequiredConstraintValidatorTest.kt:
```diff
@@ -66,4 +66,68 @@ class RequiredConstraintValidatorTest {
+  }
+
+  @Test
+  fun shouldReturnValidResultOnMultipleAnswersWithNoValue() {
```
```suggestion
  fun shouldReturnInvalidResult_noAnswerHasValue() {
```

jingtang10 commented on datacapture/src/test/java/com/google/android/fhir/datacapture/validation/RequiredConstraintValidatorTest.kt:
```diff
@@ -66,4 +66,68 @@ class RequiredConstraintValidatorTest {
     assertThat(validationResult.message).isEqualTo("Missing answer for required field.")
   }
+
+  @Test
+  fun shouldReturnInvalidResultOnAnswerWithNoValue() {
+    val questionnaireItem = Questionnaire.QuestionnaireItemComponent().apply { required = true }
+    val questionnaireResponseItem =
+      QuestionnaireResponse.QuestionnaireResponseItemComponent().apply {
+        addAnswer(QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent())
+      }
+
+    val validationResult =
+      RequiredConstraintValidator.validate(
+        questionnaireItem,
+        questionnaireResponseItem,
+        InstrumentationRegistry.getInstrumentation().context
+      )
+    assertThat(validationResult.isValid).isFalse()
+    assertThat(validationResult.message).isEqualTo("Missing answer for required field.")
+  }
```
remove this test case.

jingtang10 commented on datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemBooleanTypePickerViewHolderFactoryInstrumentedTest.kt:
```diff
@@ -322,16 +322,46 @@ class QuestionnaireItemBooleanTypePickerViewHolderFactoryInstrumentedTest {
 
-    assertThat(viewHolder.itemView.findViewById<TextView>(R.id.error_text_view).text)
-      .isEqualTo("Missing answer for required field.")
+    assertThat(viewHolder.itemView.findViewById<TextView>(R.id.error_text_view).error)
```
i think you need to assert on `.text` rather than `.error`

```suggestion
    assertThat(viewHolder.itemView.findViewById<TextView>(R.id.error_text_view).text)
```

jingtang10 commented on datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemEditTextSingleLineViewHolderFactoryInstrumentedTest.kt:
```diff
@@ -205,6 +209,50 @@ class QuestionnaireItemEditTextSingleLineViewHolderFactoryInstrumentedTest {
       .isNull()
   }
 
+  @Test
+  @UiThreadTest
+  fun displayValidationResult_answerWithNoValue_shouldShowErrorMessage() {
+    viewHolder.bind(
+      QuestionnaireItemViewItem(
+        Questionnaire.QuestionnaireItemComponent().apply {
+          required = true
+          addAnswerOption(
+            Questionnaire.QuestionnaireItemAnswerOptionComponent().apply {
+              value = StringType("hello")
+            }
+          )
+        },
+        QuestionnaireResponse.QuestionnaireResponseItemComponent().apply { addAnswer() }
+      ) {}
+    )
+
+    assertThat(viewHolder.itemView.findViewById<TextInputLayout>(R.id.text_input_layout).error)
+      .isEqualTo(
+        viewHolder.itemView.context.getString(R.string.required_constraint_validation_error_msg)
+      )
+  }
```
this is the same as `displayValidationResult_error_shouldShowErrorMessage` for the widget itself. so this can be removed.

jingtang10 commented on datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemBooleanTypePickerViewHolderFactoryInstrumentedTest.kt:
```diff
@@ -322,16 +322,46 @@ class QuestionnaireItemBooleanTypePickerViewHolderFactoryInstrumentedTest {
+      ) {}
+    )
+
+    assertThat(viewHolder.itemView.findViewById<TextView>(R.id.error_text_view).error)
```
```suggestion
    assertThat(viewHolder.itemView.findViewById<TextView>(R.id.error_text_view).text)
```

jingtang10 commented on datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemBooleanTypePickerViewHolderFactoryInstrumentedTest.kt:
```diff
@@ -322,16 +322,46 @@ class QuestionnaireItemBooleanTypePickerViewHolderFactoryInstrumentedTest {
   @UiThreadTest
-  fun displayValidationResult_error_shouldShowErrorMessage() {
+  fun displayValidationResult_questionnaireResponseAnswerWithValue_shouldShowNoErrorMessage() {
     viewHolder.bind(
       QuestionnaireItemViewItem(
-        Questionnaire.QuestionnaireItemComponent().apply { required = true },
+        Questionnaire.QuestionnaireItemComponent().apply {
+          required = true
+          addAnswerOption(
+            Questionnaire.QuestionnaireItemAnswerOptionComponent().apply { text = "Question?" }
+          )
+        },
+        QuestionnaireResponse.QuestionnaireResponseItemComponent().apply {
+          addAnswer(
+            QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
+              value = BooleanType(false)
+            }
+          )
+        }
+      ) {}
+    )
```
this is covered by `displayValidationResult_noError_shouldShowNoErrorMessage`

remove this test case

aurangzaibumer commented on datacapture/src/test/java/com/google/android/fhir/datacapture/validation/RequiredConstraintValidatorTest.kt:
```diff
@@ -66,4 +66,68 @@ class RequiredConstraintValidatorTest {
+  }
+
+  @Test
+  fun shouldReturnValidResultOnMultipleAnswersWithValue() {
```
resolved

aurangzaibumer commented:
Comments are resolved. Can you please review @jingtang10 . Thanks
cc @f-odhiambo

1371523503 commented:
什么


---Original---
From: ***@***.***&gt;
Date: Thu, Apr 7, 2022 01:10 AM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;
Subject: Re: [google/android-fhir] Fixed Showing no validation errors whenform opens (PR #1102)




 
Comments are resolved. Can you please review @jingtang10 . Thanks
 cc @f-odhiambo
 
&mdash;
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***&gt;

aurangzaibumer commented:
Unable to update with master branch @jingtang10 
@f-odhiambo

aurangzaibumer commented:
> Unable to update with master branch @jingtang10 @f-odhiambo

fixed, can update now.

jingtang10 commented on datacapture/src/androidTest/java/com/google/android/fhir/datacapture/views/QuestionnaireItemBooleanTypePickerViewHolderFactoryInstrumentedTest.kt:
```diff
@@ -322,16 +322,21 @@ class QuestionnaireItemBooleanTypePickerViewHolderFactoryInstrumentedTest {
   @Test
   @UiThreadTest
-  fun displayValidationResult_error_shouldShowErrorMessage() {
+  fun displayValidationResult_noQuestionnaireResponseAnswer_shouldShowNoErrorMessage() {
```
so you effectively removed the test case where there is an error.

jingtang10 commented:
Two possible solutions (pending investigation):

1. instead of always creating a QuestionnaireResponseItem in the QuestionnaireItemViewItem, we can leave it as `null` if the question hasn't been touched by the user at all. If this field is null (meaning there's no answer at all), we don't validate. But if it's not null, we will validate. And if the non-null QuestionnaireResponseItem has no answer, that means the user has tried to answer it but left it empty/blank. In this case if the question is required, we should show validation error.
2. Add a dirty field to indicate if the question has been touched by the user

jingtang10 commented:
@aurangzaibumer @ellykits

aurangzaibumer commented:
New PR Link https://github.com/google/android-fhir/pull/1309
